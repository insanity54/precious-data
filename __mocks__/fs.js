
const fs = jest.createMockFromModule('fs')
const fsa = jest.requireActual('fs')
const Promise = require('bluebird')
const path = require('path')

// greetz https://jestjs.io/docs/en/manual-mocks#examples

let mockFiles = Object.create(null);

function __setMockFiles(newMockFiles) {
  mockFiles = Object.create(null);
  for (const file in newMockFiles) {
    const content = newMockFiles[file]

    if (!mockFiles[file]) {
      mockFiles[file] = content;
    }

  }
}

function readFile (filePath, opts) {
  return new Promise((resolve, reject) => {
    (mockFiles[filePath]) ? resolve(mockFiles[filePath]) : reject(new Error('ENOENT'))
  })
}

const mocks = {
  __setMockFiles: __setMockFiles,
  createWriteStream: () => {
    return fsa.createWriteStream('/dev/null')
  },
  writeFile: (file, data, cb) => { return cb(null) },
  promises: {
    mkdir: (path, opts) => { return new Promise.resolve() },
    readFile: readFile,
    writeFile: (file, data) => { return new Promise.resolve() }
  }
}


// extend mocked fs with my custom mocks
for (const m in mocks) {
  fs[m] = mocks[m]
}





module.exports = fs
