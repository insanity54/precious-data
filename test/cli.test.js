const fs = require('fs');
const path = require('path');
const cliPath = path.join(__dirname, '..', 'p-data.js');
const Promise = require('bluebird');
const yargsParser = require('yargs-parser')
const pify = require('pify')
const glob = require('glob')
const spawn = require('child_process').spawn
const nock = require('nock')
jest.mock('fs')


nockBack = nock.back
nockBack.fixtures = path.join(__dirname, '..', 'fixtures')



// greets https://github.com/kentcdodds/split-guide/blob/fb4b2a2ebc1fb8c3c010c2af1318861b8bb1bb13/src/bin/index.test.js
function runCLIAndAssertFileOutput(args, cwd) {
  // const {
  //   exercisesDir = './exercises', exercisesFinalDir = './exercises-final'
  // } = yargsParser(args)
  return runCli(args, cwd)
    .then((stdout) => {
      expect(stdout).toMatchSnapshot()
    })
    .catch((e) => {
      console.error('runCLI error')
      console.error(e)
    })
}


function runCli(args = '', cwd = process.cwd()) {
  const isRelative = cwd[0] !== '/'
  if (isRelative) {
    cwd = path.resolve(__dirname, cwd)
  }

  return new Promise((resolve, reject) => {
    let stdout = ''
    let stderr = ''
    const command = path.join(__dirname, '..', 'p-data.js')
    const opts = {
      cwd: cwd
    }
    const child = spawn(command, args.split(' '), opts)

    child.on('error', error => {
      reject(error)
    })

    child.stdout.on('data', data => {
      stdout += data.toString()
    })

    child.stderr.on('data', data => {
      stderr += data.toString()
    })

    child.on('close', () => {
      if (stderr) {
        reject(stderr)
      } else {
        resolve(stdout)
      }
    })
  })
}


describe('p-data.js CLI', () => {
  describe('list subcommand', () => {
    it('should list all sets listed on p-memories.com', () => {
      return nockBack('cliList.1.json')
        .then(({ nockDone, context }) => {
          return runCLIAndAssertFileOutput('list --type setNames --json', path.resolve(__dirname, '../'))
            .then(() => {
              context.assertScopesFinished()
            })
            .then(nockDone)
        })
    })
  })
  describe('index subcommand', () => {
    it('should run without failure', () => {
      return runCLIAndAssertFileOutput('index', path.resolve(__dirname, '../'))
    })
    it('should create a setAbbrIndex.json file', () => {
      require('fs').__setMock
      return runCli('index', path.resolve(__dirname, '../'))
        .then((stdout) => {

        })
    })
  });
  describe('rip subcommand', () => {
    it(
      'should make a Set suggestion if the Set abbreviation doesnt have an exact match',
      () => {
        const parser = require('yargs')().command(require('../commands/rip'))
        return new Promise((resolve, reject) => {
          parser.parse('rip -s OREIMO', (err, argv, output) => {
            if (err) reject(err)
            resolve(output)
          })
        }).then((output) => {
          expect(output).toMatch(/Did you mean/)
        });

      }
    )
    it('should rip a card given a URL to the card', () => {
      return execFile(cliPath, ['node', './p-data.js', 'rip', '-u', 'http://p-memories.com/node/906300'])
        .then((res) => {
          expect(typeof res).toBe('string');
          expect(res).toMatch(/涼風 青葉/)
        });
    })
    it(
      'should rip a card given the set abbreviation and card number',
      () => {
        return execFile(cliPath, ['node', './p-data.js', 'rip', '-s', 'SG', '-n', '01-019'])
          .then((res) => {
            expect(typeof res).toBe('string');
            expect(res).toMatch(/岡部 倫太郎/);
          });
      }
    )
    it('should rip an entire set given the set abbreviation', () => {
      let program = cli.parse(['node', './p-data.js', 'rip', '--set', 'MMDK']);
      expect(program).toBeDefined();
      expect(program.set).toEqual('MMDK');

      // .then((res) => {
      //   assert.isString(res);
      //   assert.match(res, /劇場版 魔法少女まどか☆マギカ/);
      //   const
      // })
    })
    xit('should have silent output when called with -q flag', function() {

    })

    it('should rip all incrementally with a throttle', () => {
      let program = cli.parse(['node', './p-data.js', 'rip', '-a', '-i', '-t', '3']);
      expect(program).toBeDefined();
      console.log(program);
    })
  })
})
