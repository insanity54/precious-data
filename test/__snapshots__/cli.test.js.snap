// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p-data.js CLI list subcommand should list all sets listed on p-memories.com 1`] = `
"[\\"ダンジョンに出会いを求めるのは間違っているだろうかⅡ\\",\\"劇場版 ダンジョンに出会いを求めるのは間違っているだろうか ―オリオンの矢―\\",\\"初音ミク\\",\\"傷物語\\",\\"<物語>シリーズ\\",\\"<物語>シリーズ セカンドシーズン\\",\\"偽物語\\",\\"化物語\\",\\"とある科学の超電磁砲T\\",\\"とある科学の超電磁砲＆とある科学の超電磁砲Ｓ\\",\\"ネコぱら\\",\\"私、能力は平均値でって言ったよね！\\",\\"俺を好きなのはお前だけかよ\\",\\"ましろ色シンフォニー\\",\\"ゆるゆり、\\",\\"ゆるゆり　なちゅやちゅみ！\\",\\"ゆるゆり♪♪\\",\\"ゆるゆり\\",\\"ぼくたちは勉強ができない！\\",\\"劇場版「SHIROBAKO」\\",\\"SHIROBAKO\\",\\"まちカドまぞく\\",\\"ゆるキャン△\\",\\"ガールズ＆パンツァー 最終章\\",\\"ガールズ＆パンツァー　劇場版\\",\\"ガールズ＆パンツァー\\",\\"五等分の花嫁\\",\\"劇場版 ハイスクール・フリート\\",\\"ハイスクール・フリート\\",\\"グランベルム\\",\\"結城友奈は勇者である -鷲尾須美の章-/-勇者の章-\\",\\"結城友奈は勇者である\\",\\"かぐや様は告らせたい～天才たちの恋愛頭脳戦～\\",\\"がっこうぐらし！\\",\\"八月のシンデレラナイン\\",\\"Aチャンネル\\",\\"ひだまりスケッチ\\",\\"ひだまりスケッチ×ハニカム\\",\\"ハナヤマタ\\",\\"ケムリクサ\\",\\"ゾンビランドサガ\\",\\"RELEASE THE SPYCE\\",\\"SSSS.GRIDMAN\\",\\"劇場版 魔法少女まどか☆マギカ[新編]叛逆の物語\\",\\"劇場版 魔法少女まどか☆マギカ\\",\\"魔法少女まどか☆マギカ\\",\\"はたらく細胞\\",\\"ポプテピピック\\",\\"STEINS;GATE\\",\\"ハイスクールＤ×Ｄ HERO\\",\\"ハイスクールＤ×Ｄ ＢｏｒＮ\\",\\"とらドラ！\\",\\"怪獣娘（黒）\\",\\"エロマンガ先生\\",\\"ノーゲーム・ノーライフ ゼロ\\",\\"フレームアームズ・ガール\\",\\"NEW GAME!!\\",\\"NEW GAME!\\",\\"劇場版 魔法科高校の劣等生 星を呼ぶ少女\\",\\"やはり俺の青春ラブコメはまちがっている。続\\",\\"やはり俺の青春ラブコメはまちがっている。\\",\\"ハンドシェイカー\\",\\"きんいろモザイク Pretty Days\\",\\"ハロー！！きんいろモザイク\\",\\"きんいろモザイク\\",\\"学戦都市アスタリスク\\",\\"映画「けいおん！」\\",\\"けいおん！！Part.2\\",\\"けいおん！！\\",\\"けいおん！Part2\\",\\"けいおん！\\",\\"WORKING!!!\\",\\"WORKING´!!\\",\\"WORKING！！\\",\\"俺の妹がこんなに可愛いわけがない。\\",\\"俺の妹がこんなに可愛いわけがない\\",\\"ささみさん＠がんばらない\\",\\"咲-Saki- 全国編\\",\\"恋愛ラボ\\",\\"あの日見た花の名前を僕達はまだ知らない。\\",\\"冴えない彼女の育てかた\\",\\"カードキャプターさくら\\",\\"百花繚乱 サムライブライド\\",\\"百花繚乱 サムライガールズ\\",\\"アマガミSS+ plus\\",\\"みなみけ\\",\\"俺の彼女と幼なじみが修羅場すぎる\\",\\"さくら荘のペットな彼女\\",\\"魔法使いの夜\\",\\"恋と選挙とチョコレート\\",\\"ココロコネクト\\",\\"生徒会の一存 Lv.2\\",\\"AKB0048\\",\\"夏色キセキ\\",\\"侵略!?イカ娘\\",\\"侵略！イカ娘\\",\\"神様のメモ帳\\",\\"カーニバル・ファンタズム\\",\\"まよチキ！\\",\\"電波女と青春男\\",\\"オオカミさんと七人の仲間たち\\",\\"刀語\\",\\"ClariS\\"]
"
`;
